framework/aster-frame/src/arch/riscv/mm.rs

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::ops::Range;
+
+use alloc::fmt;
+
+use pod::Pod;
+
+use crate::vm::{
+    page_prop::{CachePolicy, PageFlags, PageProperty, PrivilegedPageFlags as PrivFlags},
+    page_table::PageTableEntryTrait,
+    Paddr, PagingConstsTrait, Vaddr, PAGE_SIZE,
+};
+
+pub(crate) const NR_ENTRIES_PER_PAGE: usize = 512;
+
+#[derive(Debug)]
+pub struct PagingConsts {}
+
+impl PagingConstsTrait for PagingConsts {
+    const BASE_PAGE_SIZE: usize = 4096;
+    const NR_LEVELS: usize = 4;
+    const HIGHEST_TRANSLATION_LEVEL: usize = 4;
+    const PTE_SIZE: usize = core::mem::size_of::<PageTableEntry>();
+}
+
+bitflags::bitflags! {
+    #[derive(Pod)]
+    #[repr(C)]
+    /// Possible flags for a page table entry.
+    pub struct PageTableFlags: usize {
+        /// Specifies whether the mapped frame or page table is valid.
+        const VALID =           1 << 0;
+        /// Controls whether reads to the mapped frames are allowed.
+        const READABLE =        1 << 1;
+        /// Controls whether writes to the mapped frames are allowed.
+        const WRITABLE =        1 << 2;
+        /// Controls whether execution code in the mapped frames are allowed.
+        const EXECUTABLE =      1 << 3;
+        /// Controls whether accesses from userspace (i.e. U-mode) are permitted.
+        const USER =            1 << 4;
+        /// Indicates that the mapping is present in all address spaces, so it isn't flushed from
+        /// the TLB on an address space switch.
+        const GLOBAL =          1 << 5;
+        /// Whether the memory area represented by this entry is accessed.
+        const ACCESSED =        1 << 6;
+        /// Whether the memory area represented by this entry is modified.
+        const DIRTY =           1 << 7;
+        // PBMT: Non-cacheable, idempotent, weakly-ordered (RVWMO), main memory
+        const PBMT_NC =         1 << 61;
+        // PBMT: Non-cacheable, non-idempotent, strongly-ordered (I/O ordering), I/O
+        const PBMT_IO =         1 << 62;
+        /// Naturally aligned power-of-2
+        const NAPOT =           1 << 63;
+    }
+}
+
+pub(crate) fn tlb_flush_addr(vaddr: Vaddr) {
+    unsafe {
+        riscv::asm::sfence_vma(0, vaddr);
+    }
+}
+
+pub(crate) fn tlb_flush_addr_range(range: &Range<Vaddr>) {
+    for vaddr in range.clone().step_by(PAGE_SIZE) {
+        tlb_flush_addr(vaddr);
+    }
+}
+
+pub(crate) fn tlb_flush_all_excluding_global() {
+    // TODO: excluding global?
```

Comment:
[RISCV privileged spec](https://uim.fei.stuba.sk/wp-content/uploads/2018/02/riscv-privileged-2022.pdf) page 85 says:

> Global mappings need not be stored redundantly in address-translation caches for multiple ASIDs. Additionally, they need not be flushed from local address-translation caches when an SFENCE.VMA instruction is executed with rs2 != x0.

Reply:
RISC-V privileged spec says

> The fence also invalidates all address-translation cache entries, for all address spaces

That is, `sfence.vma` is `tlb_flush_all_including_global`. So the problem is that, I am not sure if RISC-V has a mechanism that can cheaply do `tlb_flush_all_excluding_global`.

It seems that the reason why `tlb_flush_all_excluding_global` was added is because in x86, `tlb_flush_all_excluding_global` is a simple `Cr4` write, while `tlb_flush_all_including_global` takes a little more effort. So if we don't have the scenario that needs to flush all TLB and strictly excluding global, we can safely fallback `tlb_flush_all_excluding_global` to `tlb_flush_all_including_global`.